Changing SAM device naming scheme, add more SAM devices

chris-durand opened the pull request:
Change device naming scheme to allow for useful grouping according to logical device families:
```
    SAM      D      21      E    15       A    -    M       U
{platform}{    series    }{pin}{flash}{variant}-{package}{grade}
instead of
{platform}{family}{series}{pin}{flash}{variant}-{package}{grade}
```

The device family letter is merged into the series part because related devices might not have the same name prefix (e.g. `SAMD51` and `SAME54`).

More devices closely related to those already supported in modm are added.

- [x] Add SAME51, SAME53 and SAME54 devices
- [x] Add SAME70, SAMS70, and SAMV71 devices
- [x] Add SAMD09, SAMD10, SAMD11, SAMD20, SAMDA1 devices
- [x] Remove `family` from identifier naming scheme
- [x] Fix incorrect parsing of package codes with two letters
- [x] Combine some device files from same family

Updated device families:
SAMD51, SAME51, SAMD53, SAMD54: `D5x/E5x`
SAMD09, SAMD10, SAMD11, SAMD20, SAMD21, SAMDA1: `D1x/D2x/DAx`
SAMG5*: `G5x`
SAME70, SAMS70, SAMV70, SAMV71: `E7x/S7x/V7x`

salkinium commented on tools/generator/dfg/sam/sam_groups.py:
```diff
@@ -11,6 +11,9 @@
         'series': ['51']
+    },{
+        'family': ['e'],
+        'series': ['51', '53', '54']
```
The `sam_generator.py` isn't accessing this information, since I guess it wasn't necessary until now: This is how it works for STM32: https://github.com/modm-io/modm-devices/blob/develop/tools/generator/stm_generator.py#L40

salkinium replied:
I'm actually not sure if that is all that's missing. Seems to be another issue?

chris-durand replied:
Doesn't the merging happen [here](https://github.com/modm-io/modm-devices/blob/8916317461c65855b13582be7483daeccf65fbac/tools/generator/dfg/generator.py#L21) and should also work with SAM? The group definitions are passed to that function. I didn't get to debugging it so far.

chris-durand replied:
The device nomenclature for SAM devices is far more confusing than for STM. D51, E51, E53, and E54 series form one device family with a shared manual. A D51 is basically a E54 without Ethernet and CAN.
The second character seems to indicate the CPU core: 1: CM23, 2,A: CM0+, 4,5: CM4, 7:CM7. A D21 which is a CM0+ at 48 MHz is not at all closely related to a D51. But E70/S70/V70/V71 are all very similar although they use different letters.

I guess we have to go through some of the manuals to see which series actually share similar peripherals. Currently a lot of devices have the device types set to the name of the series, e.g. "samg51" has a uart of type "samg51" and for "samg53" it's "samg53". Probably many of them will turn out to be identical hardware.

salkinium replied:
Dear lord…

I would just delegate the comparison to when you implement drivers in modm. It's not too difficult to accept another type in modm, but it's difficult to exclude a accidentally common type. So solve this properly someone™ should compare the MMIO registers algorithmically to find shared peripherals.

chris-durand replied:
@salkinium The external memory controller shows up in the memory list as:
```
<memory name="extram" access="rwx" start="0x04000000" size="16777216"/>
```
The core module will try to use it for heap because it's the largest memory containing the string "ram" in it's name. For STM32 the external memory controller does not show up there in the device file. Should I filter it in the core module or change something in the device file?

salkinium replied:
This should be removed from the Device File, since the extram is not part of the device, similiar to FSMC on STM32.

salkinium reviewed (approved):
This is fine with me, but I'll merge together with the modm PR.

chris-durand committed:
[dfg] Add SAME5x devices

chris-durand committed:
[dfg] Add complete SAMx7x device family

chris-durand commented:
Also added more SAMx7x devices.

salkinium commented on README.md:
```diff
@@ -12,21 +12,21 @@ These tools and this data set is maintained and curated by
-| STM32L0       |   354   | STM32L1       |   180   | STM32L4       |   417   |
-| STM32L5       |    44   | STM32U5       |    61   | STM32WB       |    37   |
-| STM32WL       |    27   |
+| NRF52         |     7   | RP20          |     1   | S             |    48   |
```
S? probably `family` related issues?

chris-durand committed:
[dfg] Add SAM GCLK data

chris-durand committed:
[dfg] Change SAM identifier scheme, fix incorrect identifier parsing

SAM devices with the same letter prefix in the identifier are not
necessarily similar or related. For instance, D51 and E51 are closely
related but D21 and D51 devices are not. Therefore, the naming scheme is
changed to not contain a device family prefix anymore.

The new naming scheme is defined as follows:
    SAM     D21     E    15       A    -    M       U
{platform}{series}{pin}{flash}{variant}-{package}{grade}

The device family is now inferred from the series part of the
identifier, e.g. for D51, E51, E53 and E54 the family is "D5x/E5x".

Furthermore, identifiers with two-letter package codes are now parsed
correctly.

chris-durand committed:
[dfg] Correctly group SAM devices in Makefile, add D09, D1x, L22 devices

chris-durand committed:
[dfg] Remove duplicate GPIO data for SAMx7x revision A

chris-durand committed:
[dfg] Remove overly specific driver types for SAMV7x, SAMG5x

chris-durand committed:
[tools] Fix docs sync if device family is not part of the identifier

chris-durand committed:
Update SAM device files

chris-durand commented:
This PR has been extended to address the issue of confusing device families that do not actually represent how device are related. This is not ready to merge until modm has been adapted to the changes.

chris-durand commented on devices/sam/same70_s70_v70_v71.xml:
```diff
@@ -54,19 +168,26 @@
       <vector position="28" name="TC5"/>
       <vector position="29" name="AFEC0"/>
-      <vector position="30" name="DACC"/>
+      <vector device-series="e70|s70" device-pin="n" device-flash="21" device-variant="a" position="30" name="DACC"/>
+      <vector device-series="e70|s70|v70" device-pin="j|n|q" device-flash="19|20" device-variant="a|b" position="30" name="DACC"/>
+      <vector device-series="e70|s70|v71" device-pin="j|n|q" device-flash="19|20|21" device-variant="b" position="30" name="DACC"/>
+      <vector device-series="e70|s70|v71" device-pin="j|q" device-flash="20|21" device-variant="a" position="30" name="DACC"/>
+      <vector device-series="v71" device-pin="n|q" device-flash="19" device-variant="a" position="30" name="DACC"/>
```
That data looks quite strange but some devices of revision A really seem to have no DAC. This is contradictory to all datasheets and errata documents I could find, even the old ATMEL ones. But the device headers agree.

For SAMV71N rev A only the smallest flash size has a DAC, for SAMV71J rev A it is the other way around :confused:

chris-durand commented on tools/generator/dfg/sam/sam_device_tree.py:
```diff
@@ -159,18 +183,13 @@ def _device_tree_from_properties(p):
 
         def driver_compatibility(id):
             s = id.string
-            if s.startswith("samg55"):
-                return "samg55"
-            elif s.startswith("samg54"):
-                return "samg54"
-            elif s.startswith("samg53"):
-                return "samg53"
-            elif s.startswith("samg51"):
-                return "samg51"
-            elif s.startswith("samv70"):
-                return "samv70"
-            elif s.startswith("samv71"):
-                return "samv71"
+            # TODO: do we really need this distinction?
+            # There are two groups of devices with common peripherals:
+            # - SAM x7x,G5x
+            # - SAM D09,D1x,D2x,L2x,D51,E5x
+            # TODO: What would be the appropriate naming for those groups?
+            if s.startswith("samg5"):
+                return "samg"
             else:
                 return "sam"
```
I am not entirely sure yet what to do here. There are two separate groups of devices which have nothing in common on a register level (ignoring some IPs Atmel bought from someone else). It would be useful to be able to name them but I can't come up with anything reasonable. Any ideas?

salkinium replied:
We could name them Samelle and Samien, which are the next names on the baby name generator right after Sam. /s

Otherwise I would just call them `samg` and `samd`. A proper solution would need to look at all register maps and come up with a complete partitioning, so that you can see all variants and then make an informed decision on their names. Anything short of that is basically just lucky guessing.

chris-durand replied:
> Otherwise I would just call them `samg` and `samd`.

Probably the only sensible way of handling the peripheral types is doing that on a per peripherals basis like for STM32. If a certain type of UART peripheral is called `sercom` by Atmel we should use that naming instead of calling it a `samd` UART.

I have indeed looked at the register maps for many devices. For SAM it is a bit easier than for STM32. All the hardware IPs are reasonably labeled with id designators in the device data. If you find 5 different types of ADCs among all SAM devices you only have to compare those 5 to each other to find out which ones are similar enough to use the same driver.

We don't have to get all of it right from the start and can add additional refinement later on when we actually implement the device drivers in modm. We should nevertheless change the current type designators because labeling the peripherals as "samv71" for V71 devices and "samv70" for V70 makes really no sense when the hardware is 100% identical.

salkinium commented on tools/generator/dfg/sam/sam_identifier.py:
```diff
@@ -15,30 +15,58 @@ class SAMIdentifier:
                 i.set("package", match.group("package"))
                 i.set("grade", match.group("grade"))
+
+                i.set("family", SAMIdentifier.family_from_series(match.group("series")))
```
```suggestion
                i.set("family", SAMIdentifier.family_from_series(match.group("series").lower()))
```

chris-durand replied:
Isn't `match.group("series")` already lower case?

salkinium replied:
Oh, wrong brace, should be on the return value of `family_from_series`.

chris-durand replied:
Now lower-case. I had to put a hack into the docs sync script to correctly print `D5x/E5x` instead of capitalizing to `D5x/e5x`.

salkinium commented:
Not trying to pour oil into the fire, but looking the the family aliases you created all ending with `x`, it seems to me like the family is the series letter plus the first series number?

```
    SAM      D2      1      E    15       A    -    M       U
{platform}{    series    }{pin}{flash}{variant}-{package}{grade}
instead of
{platform}{family}{series}{pin}{flash}{variant}-{package}{grade}
```

chris-durand commented:
> Not trying to pour oil into the fire, but looking the the family aliases you created all ending with `x`, it seems to me like the family is the series letter plus the first series number?

Not really. For instance, there are the "E70, S70, V70, V71" and "D51, E51, E53, E54" families. S70 are exactly the same as E70, but without CAN and Ethernet. x7x is basically the SAM version of STM32F7 and D5x, E5x their STM32F4 series. I would not want S70 and E70 to be part of different families.

We could get rid of the `x` character in the identifier, though, if you prefer that.

salkinium commented:
Oh, that's just terrible product naming.

chris-durand commented:
@salkinium The family names containing `/` make the modm homepage generation explode since slashes in filenames are generally a bad idea:
```
FileNotFoundError: [Errno 2] No such file or directory: '/__w/modm/modm/docs/src/reference/module/modm-platform-clockgen-sam-e7x/s7x/v7x-g5x.md'
```

What would be your suggestion to handle this? My only idea right now would be to rename the families. Something like `d5x_e5x` instead of `D5x/E5x` would work but we have to then hack the display of SAM family names in the README generation ...

salkinium commented:
Just replace the / with - in [that script](https://github.com/modm-io/modm/blob/develop/tools/scripts/generate_module_docs.py#L163)?

chris-durand commented:
> Just replace the / with - in that script?

Thanks, that should work. By far the easiest fix.

salkinium reviewed (commented):
Please rebase and update the submodule in modm. Bump minor version?

chris-durand committed:
Bump version to 0.9.3
